### DISCUSSION TITLE
Roslyn code generator referenced via a 3rd package doesn't do anything

### DISCUSSION THREAD

**Discussion by paule96** [dotnet/roslyn#75132] [Q&A] [answered]
Hi roslyn community,

I build currently a bigger solution in a mono repo style, quite sucessfull. That means also we use a lot of project references instead of using nuget packages.

The problem that I'm currently facing is the following project structure:

- Project A (Some base libary project)
  - Project B (Some generator project)
- Project C (Some main project)

I let Project A reference my generator project with the following code:

```xml
<ProjectReference Include="projectb.csproj"  OutputItemType="Analyzer" ReferenceOutputAssembly="True"/>
```

And Project C referenz Project A with:

```xml
<ProjectReference Include="projecta.csproj" IncludeAssets="all" PrivateAssets="all"/>
```

The question is: How can I use my generator (Project B) in Project C with just the package reference to Project A?
Is this even possible? I understand over a nuget package reference, it is possible because you can put the generator on a special location.

**Comment by 333fred** [comment]
Well, I'll start by saying that I don't understand your project structure here. You have A referencing C, and C referencing A. That cannot work; you cannot have circular references like this, regardless of whether one is an analyzer. Can you clarify what reference structure you would like to have? Diagrams work well for this type of discussion.

**Reply by paule96** [reply]
Oh shit you where right I fixed the sample ... Their was a typo.

**Reply by 333fred** [reply]
Alright, thanks. A couple of points:

1. You shouldn't be referencing the output assembly of B in A; you want to use it as an analyzer, not as a library.
2. Analyzer assemblies _do not_ automatically flow to dependencies. You probably just want to add B as a dependency to C as well.

**Reply by paule96** [reply]
@333fred not automatically means their is a manually way to do it? It would be just nice to define some things in a normal classlib project (so the restrictions of a generator doesn't exist) and this class lib to reference the generator and provide it as an anlyser to all the projects that reference this class lib.

**Reply by 333fred** [reply]
There is with package references. I am not sure about project references, sorry.

**Comment by paule96** [comment, marked as answer]
So I kinda find a solution / workaround for me. Maybe it helps someone else.

What I do right now is the following:

### Pack your generator

Edit your Property group in your csproj of the generator like the following:

```xml
<TargetFramework>netstandard2.0</TargetFramework>
<EnforceExtendedAnalyzerRules>true</EnforceExtendedAnalyzerRules>
<NoWarn>NU5128</NoWarn>
<LangVersion>8.0</LangVersion>
<Nullable>disable</Nullable>
<ImplicitUsings>disable</ImplicitUsings>
<IsRoslynComponent>true</IsRoslynComponent>
<GeneratePackageOnBuild>true</GeneratePackageOnBuild> <!-- Generates a package at build -->
<IncludeBuildOutput>false</IncludeBuildOutput> <!-- Do not include the generator as a lib dependency -->
```

The lines about `GeneratePackageOnBuild` and `IncludeBuildOutput` are the important ones. If your solution configuration is to handle warnings as errors you need to add the `NoWarn`.
These lines will do that always if your generator is build, a nuget package will be packed, but the package will be empty for now.

### Add the generator DLL to the package

> Note: For a more percise documentation of the following, look into [the cookbook](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md#package-a-generator-as-a-nuget-package)

```xml
<ItemGroup>
  <None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" PackagePath="analyzers/dotnet/cs" Visible="false" />
</ItemGroup>
```

This will add the generator to the nuget path `analyzers/dotnet/cs`.

### Move the nuget package to your solution root

For easy handling of the generators in your monorepo style solution you should move your nuget packages to the solution root.

```xml
 <Target Name="CopyPackage" AfterTargets="Pack">
   <Copy SourceFiles="$(OutputPath)..\$(PackageId).$(PackageVersion).nupkg" DestinationFolder="$(GitRootDirectory)/packages/" />
 </Target>
```

This will copy your nuget packages to the root of your repository into the packages folder following the [.NET project structure](https://gist.github.com/davidfowl/ed7564297c61fe9ab814)

> Note: To set `GitRootDirectory` I would recommend the stackoverflow anser [here](https://stackoverflow.com/a/78907648)

### Add a nuget config

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
	<packageSources>
		<!--To inherit the global NuGet package sources remove the <clear/> line below -->
		<clear />
		<add key="nuget" value="https://api.nuget.org/v3/index.json" />
		<!--Paths for generators packages, that is needed to get reference trees for generators.-->
		<add key="YourCompany.Nuget.Packages" value="./packages" />
	</packageSources>
</configuration>
```

This Nuget config will add your solution folder as a source of Nuget packages. So your generators can be found

### Add the generator as package ref

In your project A you can now add the generator as package reference like that:

```xml
<PackageReference Include="ProjectB" Version="1.0.0" />
```

### Build order

The last step is to adjust the build order. So project A always requires that Project B was build.
For that you need to right click in your visual studio on the project A and find `Build Dependencies`.
In the sub menu you need to open `Project dependencies`.
In the new window you select the generator project.

### cleanup Nuget cache folders

~~This target will find the folder for these local produced nuget packages inside the nuget cache folders and delete it. So for each build the correct version of the generator nuget package will be used.~~

That is currently as far as I see only manual possible. Didn't found a good solution for it that runs stable

## Conclusion

That's not an easy way but it looks for now like the only solution.
The problem is that changes on the generator are not anymore applied live. You need clear the nuget cache that keeps the ref to the nuget package in your projects on the generator.
Alternativ would be some fancy versioning of the generator nuget package


Thanks @333fred for the help :)
